### PR TITLE
add possibility to disable receiver auto response

### PR DESCRIPTION
**Pull request by samyy321**
This PR is a following of the work done in https://github.com/linxGnu/gosmpp/pull/96.

It allows user to set a custom PDU as response during a `OnPDU` by setting the `DisablePDUAutoResponse` and using the `SetResponse` method of the PDU.
It also add a helper (`GetResponseWithStatus`) to create a custom response when client receive a `deliver_sm`.

### CODE REVIEW

File: pdu/PDU.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ type PDU interface {
 	// GetResponse PDU.
 	GetResponse() PDU
 
+	GetSettedResponse() *PDU
```
Comment (linxGnu) [inline pdu/PDU.go:24]: How about renaming to `GetCustomResponse()`

File: pdu/PDU.go [hunk 2]
```diff
@@ -154,6 +157,14 @@ func (c *base) IsGNack() bool {
 	return c.CommandID == data.GENERIC_NACK
 }
 
+func (c *base) SetResponse(pdu *PDU) {
```
Comment (linxGnu) [inline pdu/PDU.go:160]: Renaming to `SetCustomResponse`

File: pdu/PDU.go [hunk 3]
```diff
@@ -154,6 +157,14 @@ func (c *base) IsGNack() bool {
...
+	c.response = pdu
+}
+
+func (c *base) GetSettedResponse() *PDU {
```
Comment (linxGnu) [inline pdu/PDU.go:164]: `GetCustomResponse`

**Review by linxGnu — CHANGES_REQUESTED** [review]
I am sorry for late reviewing due to personal matter. Thank you for contributing.

I leave some comments, could you please address them.

File: pdu/PDU.go (hunk 1 above)
Comment (samyy321) [reply pdu/PDU.go]: Agree done in 46a3cb246616208c70588a33dd0b2f560b355f49

File: receivable.go [hunk 4]
```diff
@@ -126,13 +126,21 @@ func (t *receivable) handleOrClose(p pdu.PDU) (closing bool) {
...
 			if t.settings.OnPDU != nil {
 				t.settings.OnPDU(p, responded)
+				if p.CanResponse() && responded == false {
+					response := p.GetCustomResponse()
+					if response != nil {
+						t.settings.response(*response)
+					} else {
+						t.settings.response(p.GetResponse())
+					}
+				}
```
Comment (linxGnu) [inline receivable.go:143]: I think what do you want here is custom response, and you prefer it.

In this way, you don't need `DisablePDUAutoResponse`.

Just change the source code into:
```go
if p.CanResponse() && t.settings.response != nil {
   // Change internal logic
   response := p.GetCustomResponse()
   if response != nil {
        t.settings.response(*response)
   } else {
       t.settings.response(p.GetResponse())
   }
   responded = true
}

// Keep this
if t.settings.OnPDU != nil {
    t.settings.OnPDU(p, responded)
}
```

Comment (linxGnu) [reply receivable.go:143]: As you can see, we don't need dangerous setting `DisablePDUAutoResponse`

All we need to do is changing the logic of how we get the response.

**Comment by linxGnu** [comment]
Please also update your branch with origin/master @samyy321

File: receivable.go [hunk 5]
```diff
@@ -126,13 +126,21 @@ func (t *receivable) handleOrClose(p pdu.PDU) (closing bool) {
...
 				t.settings.OnPDU(p, responded)
+				if p.CanResponse() && !responded {
+					response := p.GetCustomResponse()
+					if response != nil {
+						t.settings.response(*response)
+					} else {
+						t.settings.response(p.GetResponse())
+					}
+				}
 			}
```
Comment (goten4) [inline receivable.go:144]:
```suggestion
			if p.CanResponse() && t.settings.response != nil {
				responded = true
			}

			if t.settings.OnPDU != nil {
				t.settings.OnPDU(p, responded)
			}

			if responded {
				response := p.GetCustomResponse()
				if response == nil {
					response = p.GetResponse()
				}
				t.settings.response(response)
			}
```

File: transceivable.go [hunk 6]
```diff
@@ -67,6 +67,8 @@ func newTransceivable(conn *Connection, settings Settings) *transceivable {
 			}
 		},
 
+		DisablePDUAutoResponse: settings.DisablePDUAutoResponse,
```
Comment (goten4) [inline transceivable.go:70]: No need

**Review by goten4 — CHANGES_REQUESTED** [review]
@linxGnu : @samyy321 is right, in this case we need to set the custom response in the `OnPDU()` callback, so the method `t.settings.response()` should be called after the `t.settings.OnPDU()`.

I made little suggestions without the use of `DisablePDUAutoResponse` boolean.
WDYT ?

**Comment by linxGnu** [comment]
@goten4 @samyy321 

My idea for this requirement is adding hook for custom message and prefer it over auto responding.

Something like this:
```go
type Settings struct {
	 ...
         // Hook for custom PDU response.
	CustomRespond func(pdu.PDU) (disableAutoResponse bool, response pdu.PDU)
         ...
```

```go
var responded bool
if p.CanResponse() && t.settings.response != nil {
           var disableAutoResponse bool
           var customResponse pdu.PDU

           if t.settings.CustomRespond != nil {
                 disableAutoResponse, customResponse  = t.settings.CustomRespond(p)
                 if customResponse != nil {
                      t.settings.response(customResponse)
                      responded = true
                 } 
           } 

           if !responded && !disableAutoResponse {
                  t.settings.response(p.GetResponse())
                  responded = true
           }
}

if t.settings.OnPDU != nil {
      t.settings.OnPDU(p, responded)
}
```

**Comment by goten4** [comment]
> @goten4 @samyy321
> 
> My idea for this requirement is adding hook for custom message and prefer it over auto responding.
> 
> Something like this:
> 
> ```go
> type Settings struct {
> 	 ...
>          // Hook for custom PDU response.
> 	CustomRespond func(pdu.PDU) (disableAutoResponse bool, response pdu.PDU)
>          ...
> ```
> 
> ```go
> var responded bool
> if p.CanResponse() && t.settings.response != nil {
>            var disableAutoResponse bool
>            
>            if t.settings.CustomRespond != nil {
>                   var customResponse pdu.PDU
> 
>                  disableAutoResponse, customResponse  = t.settings.CustomRespond(p)
>                  if customResponse != nil {
>                       t.settings.response(customResponse)
>                       responded = true
>                  } 
>            } 
> 
>            if !responded && !disableAutoResponse {
>                   t.settings.response(p.GetResponse())
>                   responded = true
>            }
> }
> 
> if t.settings.OnPDU != nil {
>       t.settings.OnPDU(p, responded)
> }
> ```
> 
>     * In this way, the user hook could take fully control of how to respond
> 
>     * `OnPDU` should be the same like before, only report the final result

The original issue is still there : during the `OnPDU()`, if something went wrong with the DeliverSM, you would like to respond a `ESME_RSYSERR` (for example) instead of `ESME_ROK`.

As long as the response is sent before the call of `t.settings.OnPDU(p, responded)`, you won't be able to do this.

**Comment by goten4** [comment]
@linxGnu Why not, but in that case, we should expose a way to respond ... for now the method is private, and in a Receiver connection, you do not have access to the `Submit()` method.

Maybe by passing it to the hook ? 
```
if t.settings.SelfServiceHandling != nil {
     t.settings.SelfServiceHandling(pdu, t.settings.response)
     return
} 
```

And we should return an error if the client set both the `OnPDU()` and `SelfServiceHandling()` hooks.

**Comment by goten4** [comment]
Another solution is to provide a hook that return the status of the response :
```
OnPDUWithCustomStatus func(pdu pdu.PDU) data.CommandStatusType
```

Much easier for the client of the lib ... but not for the lib.

This feature is tricky, that's why I gave up long time ago to respond a custom status when I have an error during the `OnPDU()`. I just log that I couldn't manage the PDU.

**Comment by linxGnu** [comment]
@goten4 @samyy321 

Yeah, it's better to go with Option 1, no problem if we expose Submit func. Let's give user fully control how to response.

- Most of user use transceiver, which has both Submit and Receive.
- If receiver only, we just need to expose a Response method rather than give user the response hook

Could you please make a PR 🙏

**Comment by samyy321** [comment]
Why not keeping the possibility to fallback on default response (as done in that part => https://github.com/linxGnu/gosmpp/pull/107#discussion_r1294010434) ?
For example if the user want to customize its response just for one type of PDU he doesn't have to do it for all the other types, which is not the case here by setting the `SelfServiceHandling` if I understood correctly.

**Comment by laduchesneau** [comment]
Hi all,

I've worked on this a couple of months ago, and the only easy solution I had was to break current user experience. I've been using this [code](https://github.com/linxGnu/gosmpp/compare/master...laduchesneau:gosmpp:im_the_captain_now) in production for a couple of months. ( I have a private repo with production code, but took the necessary code out and inserted into my public repo). Fair warning, this is code that works for me and I'm sharing in case anybody wants to implement their own version.


Basically, the user has full control on all PDUs received. It's their responsibility to respond to a PDU, even Unbind and EnquireLink. Receivable will not auto respond if `OnReceivePDU` is not nil. (I renamed `OnPDU` to `OnReceivedPDU`  and renamed `handleOrClose` to `handleOrRespond`)
`func (t *receivable) handleOrRespond(p pdu.PDU) (closing bool) {
	if p != nil {
		if t.settings.OnReceivedPDU != nil {
			t.settings.OnReceivedPDU(p)
		} else if p.CanResponse() && t.settings.response != nil {
			t.settings.response(p.GetResponse())
		}
	}
	return
}`

Removed all bind type interfaces, and use a single Bind interface. All bind types implement the `transceivable` type and are allowed to Send and receive PDUs. (renamed `Submit` to `Send`). Its up to the user to make sure he is sending correct PDU through each bind type. 
`type Bind interface {
	io.Closer
	Send(pdu.PDU) error
	SystemID() string
}`

Updated example/main.go to show how I manage Sending PDU.

**Comment by linxGnu** [comment]
@samyy321 @goten4 

I feel that fallback logic is adhoc and hard to control for both user and library. At anytime, user can still use pdu method to check and get the response pdu:

```go
p.CanRespond()

p.GetRespond()
```

In this way, user has clearly mindset of what to do. 

Seem like @laduchesneau is implementing our ideas here: https://github.com/linxGnu/gosmpp/pull/110

**Pull request closed without merging**